Classes/Task/OptimizeIndexTaskAdditionalFieldProvider.php

```diff
@@ -0,0 +1,281 @@
+
+        $additionalFields['site'] = [
+            'code' => $siteSelectorField->getAvailableSitesSelector('tx_scheduler[site]', $this->site),
+            'label' => $this->languageFile . ':field_site',
+            'cshKey' => '',
+            'cshLabel' => ''
+        ];
+
+        $additionalFields['cores'] = [
+            'code' => $this->getCoreSelector(),
+            'label' => $this->languageFile . ':field_cores',
+            'cshKey' => '',
+            'cshLabel' => ''
+        ];
+
+        return $additionalFields;
+    }
+
+    protected function getCoreSelector()
+    {
+        $selectorMarkup = $this->getLanguageService()->sL($this->languageFile . ':tasks.validate.selectSiteFirst');
+
+        if (is_null($this->site)) {
+            return $selectorMarkup;
+        }
+
+        $selectorField = GeneralUtility::makeInstance(CoreSelectorField::class, /** @scrutinizer ignore-type */ $this->site);
+        $selectorField->setFormElementName('tx_scheduler[cores]');
+        $selectorField->setSelectedValues($this->task->getCoresToOptimizeIndex());
+        $selectorMarkup = $selectorField->render();
+
+        return $selectorMarkup;
+    }
+
+    /**
+     * Checks any additional data that is relevant to this task. If the task
+     * class is not relevant, the method is expected to return TRUE
+     *
+     * @param array $submittedData reference to the array containing the data submitted by the user
+     * @param SchedulerModuleController $schedulerModule reference to the calling object (Scheduler's BE module)
+     * @return bool True if validation was ok (or selected class is not relevant), FALSE otherwise
+     */
+    public function validateAdditionalFields(
+        array &$submittedData,
+        SchedulerModuleController $schedulerModule
+    )
+    {
+        $result = true;
+
+        // validate site
+        $sites = $this->siteRepository->getAvailableSites();
+        if (!array_key_exists($submittedData['site'], $sites)) {
+            $result = false;
+        }
+
+        // validate core selection
+        if (array_key_exists('cores', $submittedData)
+            && !is_array($submittedData['cores'])
+        ) {
+            $this->addMessage(
+                $this->getLanguageService()->sL($this->languageFile . ':tasks.validate.invalidCores'),
+                FlashMessage::ERROR
+            );
+            $result = false;
+        }
+        return $result;
+    }
+
+    /**
+     * Saves any additional input into the current task object if the task
+     * class matches.
+     *
+     * @param array $submittedData array containing the data submitted by the user
+     * @param AbstractTask $task reference to the current task object
+     */
+    public function saveAdditionalFields(
+        array $submittedData,
+        AbstractTask $task
+    )
+    {
+        /** @var $task OptimizeIndexTask */
+        if (!$this->isTaskInstanceofOptimizeIndexTask($task)) {
+            return;
+        }
```

Comment:
Method `isTaskInstanceofOptimizeIndexTask()` never returns a false value, instead an exception is thrown.
The return statement never reached.

Reply:
Was adapted from `Classes/Task/ReIndexTaskAdditionalFieldProvider.php` and not really cared about by me. I change it in my new file.